Adding cf-m2-build-bat to facilitate using bash scripts in conda build on win-64

owinebar opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [X] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [X] Source is from official source.
- [X] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [X] If static libraries are linked in, the license of the static library is packaged.
- [X] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [X] Build number is 0.
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [X] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [X] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/cf-m2-build-bat/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/cf-m2-build-bat/meta.yaml**:

 * ❌ The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'test', 'about'].
 * ❌ The recipe could do with some maintainers listed in the `extra/recipe-maintainers` section.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/12880202403](https://github.com/conda-forge/conda-forge-webservices/actions/runs/12880202403). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/cf-m2-build-bat/meta.yaml```) and found it was in an excellent condition.

owinebar commented:
With scripts like these I was able to build emacs-feedstock on win-64 working with the build.sh.
This cleans those up and packages them for general use, i.e. anyone wanting to use shell scripts running make, etc, under conda build, without hacking conda-build itself.

mfansler commented:
Perhaps this should be proposed as a new example recipe?

CC: @isuruf

isuruf commented:
Here's an example recipe https://github.com/conda-forge/gmp-feedstock/blob/main/recipe/bld.bat

owinebar commented:
These scripts have the advantage that the user does not need to know any details of conda build.  Further, any changes to conda build or the MSYS2 implementation won't require changes to every recipe that uses m2-* packages in the build environment.  

With respect to packages already building with the m2 environment,  have the installation conventions been discussed or decided in some forum?  Using the MSYS root directory as the installation prefix has negative consequences in that the MSYS DLL masks the bin subdirectory of the root directory.  Has there been any agreement to use the UCRT environment uniformly, as opposed to the MINGW64 environment (which also, in fact, uses UCRT)?  I ask because the git package for win-64 is just a repackaging of fit for windows, which uses the MINGW64 environment.

owinebar commented:
@conda-forge/help-c-cpp Ready for review
I'm not sure this is the right team, as there is no C or C++ code, but many projects using the make build system are C/C++ projects, so it seems like the best option.

owinebar commented:
I added an argument to the trampoline scripts to specify MSYSTEM explicitly, record PREFIX from conda-build as HOST_PREFIX, then set PREFIX to MSYSTEM_PREFIX before invoking build.sh.  This should give behavior more aligned with the MSYS2 build.

owinebar commented:
I don't understand the failure of the "staged-recipes linter/liner (pull request)" check.  It might have failed because I clicked the "update branch" (to sync with main) button after the tests started running?

owinebar commented:
> Here's an example recipe https://github.com/conda-forge/gmp-feedstock/blob/main/recipe/bld.bat

Hi, @isuruf,
It took me a while to review this example, and why it might be working.

1.  The author has manually recreated the relevant steps normally generated by conda build in bld.bat.  
2.   The author was careful to append the contents of build sh to the win-build.sh script, rather than exec'ing build.sh or "bash -c build.sh", which likely would have suffered many mysterious errors due to the duplicated msys-2.0.dll on the path.
3.  I don't understand the significance of specifying the c compiler to be GCC in conda_build_config.yaml versus using the 'm2w64_c' compiler argument.  Or why the C++ configuration option is turned off under Windows since g++ is available there as well 
4. Is the fact that the weren't any weird errors while running configure or make from win-build.sh just good luck or by design?

The last one I ask because when I was attempting to get emacs compiling by running build.sh, the GCC activation script complained about an invalid sed script, and the complaints were from some backslashes that were present in the text of the activation script not being seen by sed in the expression provided to it as a command line argument.  AFAICT, that phenomenon is from the multiple instances of the library on the path, as the sed executable is linking to a different instance of the MSYS library than the shell process executing it.
The trampoline approach I've implemented is designed to safely avoid such mysterious failures.

owinebar commented:
@mfansler , @isuruf ,
I have added working recipes for GMP and emacs, which identified some bugs that are now corrected.  They are built as part of the testing, so the win_64 build is going to be running for a while 

Any advice on how to get this reviewed?  I see it as a stop-gap measure so MSYS packages can be ported in a way that works with conda's  approach to environment activation, rather than the current simple binary repackaging.  These trampolines will simplify that process, as can be seen in comparing the example GMP recipe with the gmp-feedtock recipe.

TIA.

isuruf commented:
>  I see it as a stop-gap measure so MSYS packages can be ported in a way that works with conda's approach to environment activation, rather than the current simple binary repackaging.

What do you mean by this?

owinebar commented:
> > I see it as a stop-gap measure so MSYS packages can be ported in a way that works with conda's approach to environment activation, rather than the current simple binary repackaging.
> 
> What do you mean by this?

To my mind, the simplest resolution would be to
- rebuild the keys runtime with activation/deactivation scripts that only adds Library/usr/bin to the path if there is no msys or cygwin dll on the path
- rebuild the other msys binaries to install under [Library]/msys so that executable shadowing works for everything aside from the runtime and the executables packaged directly with it
- Ensure UCRT linked executables are installed under some uniform environment prefix, e.g. Library/mingw64, so that Msys binaries can actually see the files installed under $PREFIX/bin without relying on conda environment activation.

owinebar commented:
Accidentally clicked the wrong comment button.

owinebar commented:
> * Ensure UCRT linked executables are installed under some uniform environment prefix, e.g. Library/mingw64, so that Msys binaries can actually see the files installed under $PREFIX/bin without relying on conda environment activation.

Actually, I got myself confused on "mingw64" while trying to understand the significance of the compiler triplets ending in "mingw32", since the "mingw64" projects default to linking to UCRT as of version 12. I forgot that MSYS2 defines the MINGW64 "environment" by linking to msvcrt.dll.

So I am going to change the default MSYSTEM value for these scripts from MINGW64 to UCRT.

Lynn

carterbox commented:
Merging because the recipe meets recipe standards. It seems like most of the discussion here is about the utility of this package, but usefullness is not one of the review criteria.

owinebar commented:
Thanks!
